### PR TITLE
Make Executor and Purger classes final

### PR DESCRIPTION
**Pull request by GromNaN**
By moving this classes to a soft `@final` in 1.8, we can make them hard `final` in 2.0 and add types to the properties.

We don't expect any further implementations to be created out of this package.

Related to this outdated PR preparing a v2 #290

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -6,6 +6,10 @@ awareness about deprecated code.
 - Use of our low-overhead runtime deprecation API, details:
   https://github.com/doctrine/deprecations/
 
+# Upgrade to 1.8
+
+Executor and Purger classes are now final. You should not extend them.
```
Comment (greg0ire) [inline UPGRADE.md]: You made the `AbstractExecutor` internal, should that be mentioned as well? What's the consequence of marking a class as `internal` for the users? I think this means that:
- they should not extend it
- they cannot even type hint against it in their code

Comment (GromNaN) [reply UPGRADE.md]: Exact. I updated the UPGRADE doc.

Comment (greg0ire) [reply UPGRADE.md]: Great! Now you just have to push it to github :stuck_out_tongue:

Comment (GromNaN) [reply UPGRADE.md]: 🤦

**Review by greg0ire — APPROVED** [review]

**Review by SenseException — APPROVED** [review]

**Pull request merged**